### PR TITLE
[SPARK-20315] [SQL] Set ScalaUDF's deterministic to true 

### PR DESCRIPTION
**Pull request by gatorsmile**
### What changes were proposed in this pull request?
ScalaUDF is always assumed to deterministic, based on the previous discussion in https://github.com/apache/spark/pull/13087. However, the current master still sets it based on the children's deterministic values. 

This PR is to correct it by always setting it to true, even if the children's deterministic values are not true.

### How was this patch tested?
Added a test case.

### CODE REVIEW

**Comment by gatorsmile** [comment]
cc @cloud-fan @dongjoon-hyun

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala [hunk 1]
```diff
@@ -45,6 +45,9 @@ case class ScalaUDF(
     udfName: Option[String] = None)
   extends Expression with ImplicitCastInputTypes with NonSQLExpression {
 
+  // the user-defined functions must be deterministic.
+  final override def deterministic: Boolean = true
```
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/ScalaUDF.scala]: Oh, although we decided that this must be `true`, but should we override this explicitly? Is this for allowing further optimizations like predicate pushdown?

**Comment by viirya** [comment]
Hmm, I think it is a bit different between the deterministic assumption on UDF functions and the `deterministic` of `ScalaUDF`.

Even your UDF functions are deterministic, if the input expressions are not deterministic, the result of this ScalaUDF isn't deterministic.

**Comment by gatorsmile** [comment]
Even if the input expressions are not deterministic, the output could be still deterministic. 

If we already make an assumption that ScalaUDF is deterministic, we should make it behave consistent.

**Comment by viirya** [comment]
udf(x, y) = x + y looks like a deterministic UDF function. Is this udf(rand(), rand()) deterministic?

**Comment by gatorsmile** [comment]
A simple example, `udf(x, y) = 1` is deterministic no matter whether x or y is deterministic or not.

**Comment by gatorsmile** [comment]
The deterministic of our [Scala udaf has exactly the same logics](https://github.com/apache/spark/blob/489845f3a0e2a3555b96b6f3dbb984c783b20d97/sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/udaf.scala#L339
) and Hive UDF/UDAF... have the same logics too

**Comment by viirya** [comment]
`udf(x, y) = 1` is deterministic no matter whether x or y is deterministic or not, is because x, y are not used, in other words they don't affect the result of the udf.

The result of `udf(x, y) = x + y` is deterministic or not, depending on whether x and y are deterministic too. The result of `udf(rand(), rand())` is not deterministic, for example.

I think `ScalaUDAF` is a bit different to `ScalaUDF`, because we don't push down an aggregate function. But we will push down a `ScalaUDF` if it is deterministic. Pushing down it might cause unexpected results.

**Comment by gatorsmile** [comment]
After I rethinking about it, it really depends on how we define `deterministic`. 

I checked the definition of Hive `UDFType`. 
```
  /**
   * Certain optimizations should not be applied if UDF is not deterministic.
   * Deterministic UDF returns same result each time it is invoked with a
   * particular input. This determinism just needs to hold within the context of
   * a query.
   *
   * @return true if the UDF is deterministic
   */
  boolean deterministic() default true;
```

The deterministic value of an expression should always consider the children's deterministic values. All the other existing UDFs in Spark are not correct. Thus, I will close it and fix them in a separate PR.

**Pull request closed without merging**